### ISSUE TITLE
Recursion error in `VaultApiCategory.__getattr__()`

### ISSUE THREAD

**Issue by cansjt** [hvac/hvac#838]
We are seeing a lot of errors like this, in two of our environments.
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/hvac/api/vault_api_category.py", line 39, in __getattr__
    if item in self.implemented_class_names:
  File "/usr/local/lib/python3.10/site-packages/hvac/api/vault_api_category.py", line 39, in __getattr__
    if item in self.implemented_class_names:
  File "/usr/local/lib/python3.10/site-packages/hvac/api/vault_api_category.py", line 39, in __getattr__
    if item in self.implemented_class_names:
  [Previous line repeated 997 more times]
RecursionError: maximum recursion depth exceeded
```
The runtime context is pods running in a kubernetes context. The outcome does not seem deterministic though as some pods with the same config will fail, while others won't. Could it be related to the fact that some module aren't always loaded in the same order? Or API not listed in the same order?

Just by [reading the code](https://github.com/hvac/hvac/blob/develop/hvac/api/vault_api_category.py#L41), I would say that the only way this happens is that an api category does not have the expected private attribute define (some API impl. not respecting the implementation contract), and hence `getattr()` falls in a recursion pit.

Not sure how/where those api categories are defined yet though.

**Comment by briantist** [comment]
Hi @cansjt is this something you're still experiencing?

**Comment by m4dh4t** [comment]
I'm encountering the same error when working in a multiprocessing pool context :

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/usr/lib/python3.10/multiprocessing/queues.py", line 367, in get
    return _ForkingPickler.loads(res)
  File "*/venv/lib/python3.10/site-packages/hvac/api/vault_api_category.py", line 37, in __getattr__
    if item in self.implemented_class_names:
  File "*/venv/lib/python3.10/site-packages/hvac/api/vault_api_category.py", line 37, in __getattr__
    if item in self.implemented_class_names:
  File "*/venv/lib/python3.10/site-packages/hvac/api/vault_api_category.py", line 37, in __getattr__
    if item in self.implemented_class_names:
  [Previous line repeated 972 more times]
RecursionError: maximum recursion depth exceeded
```

**Comment by m4dh4t** [comment]
So, after a bit more digging I narrowed down the issue to the way the `pickle` library interacts with the `hvac.v1.Client` object, which means it is possible to reduce the problematic code example to:

```python
v = hvac.Client()
copy.deepcopy(v)  # RecursionError('maximum recursion depth exceeded')
```

or at a lower level:

```python
v = hvac.Client()
pickled = pickle.dumps(v)
unpickled = pickle.loads(pickled)  # RecursionError('maximum recursion depth exceeded')
```

The reason for this infinite recursion boils down to some attributes not being fully initialized when passed to the `__getattr__` function, which cannot find them and thus invokes itself in a loop. This seems to be a known bug as it already impacted _Coverage.py_ a while back, for which a more technical writeup can be found [here](https://nedbatchelder.com/blog/201010/surprising_getattr_recursion.html).

In our case, the culprits are the attributes `implemented_class_names` in `vault_api_category.py` and `default_kv_version` in `kv.py`, for which I made this [fix](https://github.com/hvac/hvac/pull/982) that seemingly does the trick from what I was able to test.
